lib/scanner/xml-scanner/xml-scanner.h

```diff
@@ -0,0 +1,93 @@
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#ifndef XML_SCANNER_H_INCLUDED
+#define XML_SCANNER_H_INCLUDED
+
+#include "syslog-ng.h"
+#include "messages.h"
+#include "logmsg/logmsg.h"
+
+#include <string.h>
+
+typedef void (*PushCurrentKeyValue)(const gchar *name, const gchar *value, gssize value_length, gpointer user_data);
+
+typedef void (*StartElement) (GMarkupParseContext *context, const gchar *element_name, const gchar **attribute_names,
+                              const gchar **attribute_values, gpointer user_data, GError **error);
+typedef void (*EndElement) (GMarkupParseContext *context, const gchar *element_name, gpointer user_data,
+                            GError **error);
+typedef void (*Text) (GMarkupParseContext *context, const gchar *text, gsize text_len, gpointer user_data,
+                      GError **error);
+
+typedef struct
+{
+  gboolean strip_whitespaces;
+  GList *exclude_tags;
+} XMLScannerOptions;
+
+typedef struct
+{
+  GMarkupParseContext *xml_ctx;
+  GMarkupParser scanner_callbacks;
+  XMLScannerOptions *options;
+  gboolean pop_next_time;
+  gboolean matchstring_shouldreverse;
+  GPtrArray *exclude_patterns;
+  PushCurrentKeyValue push_function;
+  StartElement start_element_function;
+  EndElement  end_element_function;
+  Text  text_function;
+  gpointer user_data;
+  GString *key;
+  GString *text;
+} XMLScanner;
+
+XMLScanner *xml_scanner_new(void);
+void xml_scanner_free(XMLScanner *self);
+void xml_scanner_init_instance(XMLScanner *self);
+void xml_scanner_init(XMLScanner *self, XMLScannerOptions *options, PushCurrentKeyValue push_function,
+                      gpointer user_data, gchar *key_initial);
+void xml_scanner_deinit(XMLScanner *self);
+void xml_scanner_parse(XMLScanner *self, const gchar *input, gsize input_len, GError **error);
+
+static inline void
+xml_scanner_push_current_key_value(XMLScanner *self, const gchar *name, const gchar *value, gssize value_length)
+{
+  g_assert(self->push_function); // FIXME bet way to detect programming errors?
+  self->push_function(name, value, value_length, self->user_data);
+}
+
+void xml_scanner_start_element_method(GMarkupParseContext *context, const gchar *element_name,
```

Comment:
We should hide the GMarkupParseContext as we discussed IRL.

Reply:
done